src/SourceBuild/content/test/Microsoft.DotNet.SourceBuild.SmokeTests/ArtifactsSize.cs

```diff
@@ -0,0 +1,105 @@
+using System;
+using System.IO;
+using System.Linq;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
+using System.Formats.Tar;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.DotNet.SourceBuild.SmokeTests;
+
+[Trait("Category", "SdkContent")]
+public class ArtifactsSize : SmokeTests
+{
+    private readonly string baselineFilePath = BaselineHelper.GetBaselineFilePath($"ArtifactsSizes/{Config.TargetRid}.txt");
+    private readonly string[] baselineFileContent;
+    private readonly Regex buildVersionPattern = new(@"\b\d+\.\d+\.\d+[-@](alpha|preview|rc|rtm)\.\d(\.\d+\.\d+)?\b");
+
+
+    public ArtifactsSize(ITestOutputHelper outputHelper) : base(outputHelper)
+    {
+        baselineFileContent = File.Exists(baselineFilePath) ? File.ReadAllLines(baselineFilePath) : Array.Empty<string>();
+    }
+
+    [SkippableFact(new[] { Config.SourceBuiltArtifactsPathEnv, Config.SdkTarballPathEnv, Config.TargetRidEnv }, skipOnNullOrWhiteSpace: true)]
+    public void ArtifactsSizeTest()
+    {
+        Assert.True(Directory.Exists(BaselineHelper.GetBaselineFilePath("ArtifactsSizes/")));
+        Assert.NotNull(Config.SourceBuiltArtifactsPath);
+        Assert.NotNull(Config.SdkTarballPath);
+        Assert.NotNull(Config.TargetRid);
+
+        IEnumerable<TarEntry> artifactsTarEntries = Utilities.GetTarballContent(Config.SourceBuiltArtifactsPath).Where(entry => entry.EntryType == TarEntryType.RegularFile);
+        IEnumerable<TarEntry> sdkTarEntries = Utilities.GetTarballContent(Config.SdkTarballPath).Where(entry => entry.EntryType == TarEntryType.RegularFile);
+
+        string[] tarEntries = sdkTarEntries.Concat(artifactsTarEntries)
+            .Select(entry =>
+            {
+                string modifiedPath = buildVersionPattern.Replace(entry.Name, "VERSION");
+                string result = modifiedPath.Replace(Config.TargetRid, "TARGET_RID");
+                return $"{result}: {entry.Length} bytes";
+            })
+            .OrderBy(entry => entry)
+            .ToArray();
+
+        foreach (string entry in tarEntries)
+        {
+            string tarEntryFilename = entry.Substring(0, entry.IndexOf(":")).Trim();
+            string baselineEntry = baselineFileContent?.FirstOrDefault(baselineEntry => baselineEntry.StartsWith(tarEntryFilename)) ?? "";
+
+            if (string.IsNullOrEmpty(baselineEntry))
+            {
+                LogWarningMessage($"{tarEntryFilename} does not exist in baseline. Adding it to the baseline file");
+                File.AppendAllText(baselineFilePath, $"{entry}" + Environment.NewLine);
```

Comment:
I would save all file writes until the very end of the processing. There may be a lot of files that don't exist in the baseline. So that could end up being very inefficient to read the file from disk, add a new line, save it disk, and do all that for each new file. I assume we'd also want the lines in the file to be sorted and so appending them to the end isn't going to cause that to happen.

Reply:
Addressed in [this commit](https://github.com/dotnet/installer/pull/17233/commits/55d2f807485338ee2dd14c66891f9a5092731ba9).